### ISSUE TITLE
Facelets ResourceResolver does not work when suffix mapping is used

### ISSUE THREAD

**Issue by glassfishrobot** [javaee/javaserverfaces-spec#893]
This issue is being discussed on the list under the name:

[jsr-314-open] 490-XmlViews Processing JSPX files as Facelets

The following issues are related:

[https://issues.apache.org/jira/browse/MYFACES-2628](https://issues.apache.org/jira/browse/MYFACES-2628)

[https://javaserverfaces-spec-public.dev.java.net/issues/show_bug.cgi?id=781](https://javaserverfaces-spec-public.dev.java.net/issues/show_bug.cgi?id=781)

[https://javaserverfaces.dev.java.net/issues/show_bug.cgi?id=1567](https://javaserverfaces.dev.java.net/issues/show_bug.cgi?id=1567)

The resume of the issue is this:

AS >> In what way (the ResourceResolver) does not work?
AS >> It this is broken, it definitely needs to be fixed.
AS >> We need more details on exactly what behavior is broken.

The problem happen only when suffix mapping is used. In few words,
When this mapping is used, it is required to check if a
physical resource exists, usually calling ExternalContext.getResource.
Since the algorithm described on the spec does not take into
account ResourceResolver, a http 404 not found response is sent back.
(The reference that needs to be fixed on the spec is JSF 2.0
section 7.5.2 Default ViewHandler Implementation, when it describe
the requeriments of ViewHandler.deriveViewId.)

AS >> We are using it (ResourceResolver) here. Now I am getting worried
AS >> that maybe our code is broken. ![](https://java.net/jira/images/icons/emoticons/smile.gif) Though Jason also blogged on this
AS >> topic here:
AS >> [http://blogs.steeplesoft.com/2010/05/putting-facelets-in-a-jar/](http://blogs.steeplesoft.com/2010/05/putting-facelets-in-a-jar/)
AS >> Which makes me think that this must be minimally working.

I think you are using ResourceResolver with applications that uses prefix
mapping. The hack on the blog works but by accident. Try the same example,
but without configure ResourceResolver and it will work, because we are
overriding ExternalContext.getResource, which it is used to check if a
resource exists or not.

The proposal at first view, to solve this one is add a method on
ViewDeclarationLanguageFactory called existsViewId(), so the algorithm on
ViewHandler.deriveViewId just delegate to this one and that one delegate to
different ViewDeclarationLanguage implementations. But it seems better to
move all logic from section 7.5.2 ViewHandler.deriveViewId. Here I have to
mention something I said on this discussion:

[jsr-314-open] Should the ViewDeclarationLanguage be responsible to
indicate if a viewId can be created?

On mojarra, this is the code on MultiViewHandler.getViewDeclarationLanguage:

public ViewDeclarationLanguage getViewDeclarationLanguage(FacesContext
context,
String viewId)

{ String actualViewId = derivePhysicalViewId(context, viewId, false); return vdlFactory.getViewDeclarationLanguage(actualViewId); }

Now on myfaces is this:

@Override
public ViewDeclarationLanguage getViewDeclarationLanguage(
FacesContext context, String viewId)

{ // return a suitable ViewDeclarationLanguage implementation for the given viewId return _vdlFactory.getViewDeclarationLanguage(viewId); }

The difference is subtle, but very, very important. This method is
called from many locations, but only once (from RestoreViewPhase) it is
passed the "raw" viewId.
To "derive" the physical viewId it is required to know if prefix or
suffix mapping is used and if suffix mapping is used, try to check if a
view "resource" exists or not with different derived ids with the
extensions configured ".xhtml .jsp".

In this case, MyFaces "derive" the physical viewId before call
getViewDeclarationLanguage from RestoreViewPhase. In theory, we should
call from RestoreViewPhase algorithm a method exposed by ViewHandler.
I'm not have clear the details yet.
#### Environment
Operating System: All
Platform: All
#### Affected Versions
[2.0]

**Comment by glassfishrobot** [comment]
Reported by lu4242

**Comment by glassfishrobot** [comment]
Issue-Links:
blocks
[JAVASERVERFACES_SPEC_PUBLIC-490](https://github.com/javaee/javaserverfaces-spec/issues/490)
is duplicated by
[JAVASERVERFACES-1567](https://github.com/javaee/javaserverfaces/issues/1567)

**Comment by glassfishrobot** [comment]
File: [fixResourceResolver-1.patch](https://java.net/jira/secure/attachment/13778/fixResourceResolver-1.patch)
Attached By: lu4242

**Comment by glassfishrobot** [comment]
lu4242 said:
Created an attachment (id=307)
Minimal proposal to fix ResourceResolver problem

**Comment by glassfishrobot** [comment]
lu4242 said:
Created an attachment (id=308)
Solution enhanced that delegate extension configuration to VDL

**Comment by glassfishrobot** [comment]
File: [fixResourceResolver-2.patch](https://java.net/jira/secure/attachment/13779/fixResourceResolver-2.patch)
Attached By: lu4242

**Comment by glassfishrobot** [comment]
aschwart said:
Created an attachment (id=312)
Alternate fix for 893

**Comment by glassfishrobot** [comment]
File: [alternate-fix-893.patch](https://java.net/jira/secure/attachment/13780/alternate-fix-893.patch)
Attached By: aschwart

**Comment by glassfishrobot** [comment]
aschwart said:
Woops - my previous patch missed an important part of the fix: I need to actually call
ViewHandler.viewExists(). ![](https://java.net/jira/images/icons/emoticons/smile.gif) Working on this now.

**Comment by glassfishrobot** [comment]
aschwart said:
Created an attachment (id=313)
Alternate fix for 893 Take 2

**Comment by glassfishrobot** [comment]
File: [alternate-fix-893-take2.patch](https://java.net/jira/secure/attachment/13781/alternate-fix-893-take2.patch)
Attached By: aschwart

**Comment by glassfishrobot** [comment]
aschwart said:
This patch:

[https://javaserverfaces-spec-public.dev.java.net/nonav/issues/showattachment.cgi/313/alternate-](https://javaserverfaces-spec-public.dev.java.net/nonav/issues/showattachment.cgi/313/alternate-)
fix-893-take2.patch

Is a modified version of Leonardo's patch that addresses the following requirements discussed on jsr-
314-open:

1\. The ViewHandler.getViewDeclarationLanguage() should be clear that it expects "derived" view ids.
2\. We need some way to safely derive view ids from the restore view phase. This means that we either
need to be able to use the existing ViewHandler.deriveViewId() without triggering TCK failures, or we
need a new method on ViewHandler that performs the same functionality as deriveViewId() without
requiring that the derived view id passes the existence check.
3\. We need a method on ViewDeclarationLanguage for testing view existence, eg. viewExists(). We
should use this instead of directly calling ExternalContext.getResource() when checking to see whether
a viewId corresponds to a physical view.
4\. We need some way to wrap a specific VDL (eg. the JSP VDL). The minimal solution would be to
introduce a VDL id (eg. ViewDeclarationLanguage.getId()) and define standard ids for the JSP and
Facelets VDLs.

The patch was created against the Mojarra trunk, rev 8675\. It includes the following new APIs:

*   ViewHandler.deriveLogicalViewId(): Similar to ViewHandler.deriveViewId(), but without requiring the
    final test that the physical resource exists.
*   ViewDeclarationLanguage.viewExists(): Tests whether the specified viewId exists.
*   ViewDeclarationLanguage.getId(): Returns a unique id for the VDL
*   ViewDeclarationLanguage.FACELETS_VIEW_DECLARATION_ID and JSP_VIEW_DECLARATION_ID:
    constants specifying the ids for our standard VDLs.

Noteworthy implementation changes:

*   MultiViewHandler.getViewDeclarationLanguage() no longer calls deriveViewId() - ie. it expects a
    derived viewId to be provided.
*   RestoreViewPhase.execute() now uses ViewHandler.deriveLogicalViewId() to derive the viewId.
*   MultiViewHandler.deriveViewId() now uses ViewDeclarationLanguage.viewExists() to test for physical
    resource existence.

**Comment by glassfishrobot** [comment]
aschwart said:
BTW, I verified that ant test.with.container.refresh succeeds. I haven't added any new tests, though most
of this functionality should already be covered by existing tests.

**Comment by glassfishrobot** [comment]
aschwart said:
Created an attachment (id=314)
Third rev on alternate patch

**Comment by glassfishrobot** [comment]
File: [alternate-fix-893-take3.patch](https://java.net/jira/secure/attachment/13782/alternate-fix-893-take3.patch)
Attached By: aschwart

**Comment by glassfishrobot** [comment]
aschwart said:
This version of the patch:

[https://javaserverfaces-spec-public.dev.java.net/nonav/issues/showattachment.cgi/314/alternate-fix-](https://javaserverfaces-spec-public.dev.java.net/nonav/issues/showattachment.cgi/314/alternate-fix-)
893-take3.patch

Incorporates Blake's suggestion that the default implementation of ViewDeclarationLanguage.getId() return
the fully qualified class name instead of the empty string.

**Comment by glassfishrobot** [comment]
@edburns said:
Blocks 490.

**Comment by glassfishrobot** [comment]
@edburns said:
I have applied the patch and the automated tests still run to completion successfully.

**Comment by glassfishrobot** [comment]
@edburns said:
Committed revision 8686\. Also committed FrameMaker document changes.

**Comment by glassfishrobot** [comment]
@edburns said:
EB> Leonardo, In your opinion, do we need to make any changes to
EB> NavigationHandler so that the new ViewDeclarationLanguage.viewExists()
EB> method is consulted? I would expect this might need to be done.

AS> Oh, woops - yep. ConfigurableNavigationHandler implementations
AS> should use VDL.viewExists() instead of ExternalContext.getResource()
AS> to test for view existence when determining whether implicit
AS> navigation is possible. Actually, any place where we were using
AS> ExternalContext.getResource() to check for a physical view needs to
AS> go through VDL.viewExists() now. I believe that
AS> ViewHandler.deriveViewId() and implicit navigation are the only 2
AS> places where this comes up, though I will double check that later
AS> today.

OK, I have reopened the issue and will assign it to you. Please attach
a patch to the issue and send it to dev@javaserverfaces.dev.java.net for
review. When I review the patch, I'll make the necessary changes to the
spec PDF document.

Ed

**Comment by glassfishrobot** [comment]
@edburns said:
reassign

**Comment by glassfishrobot** [comment]
@edburns said:
started

**Comment by glassfishrobot** [comment]
aschwart said:
Turns out implicit navigation is already covered since it calls ViewHandler.deriveViewId(), which calls
viewExists(). Handing back over to Ed.

**Comment by glassfishrobot** [comment]
aschwart said:
No more work is required. Marking the issue as fixed again.

**Comment by glassfishrobot** [comment]
@edburns said:
Target m6.

**Comment by glassfishrobot** [comment]
Marked as **fixed** on Monday, November 1st 2010, 6:12:17 am

**Comment by glassfishrobot** [comment]
@manfredriem said:
Closing resolved issue out

**Comment by glassfishrobot** [comment]
This issue was imported from java.net JIRA JAVASERVERFACES_SPEC_PUBLIC-893
